crypto/dilithium/pqcrystals_dilithium_ref_common/sign.c

```diff
@@ -63,73 +68,92 @@ int crypto_sign_keypair(ml_dsa_params *params, uint8_t *pk, uint8_t *sk) {
+* Name:        crypto_sign_keypair
 *
-* Description: Computes signature.
+* Description: FIPS 204: Algorithm 1 ML-DSA.KeyGen
+*              Generates public and private key.
 *
 * Arguments:   - ml_dsa_params: parameter struct
+*              - uint8_t *pk: pointer to output public key (allocated
+*                             array of CRYPTO_PUBLICKEYBYTES bytes)
+*              - uint8_t *sk: pointer to output private key (allocated
+*                             array of CRYPTO_SECRETKEYBYTES bytes)
+*
+* Returns 0 (success) -1 on failure
+**************************************************/
+int crypto_sign_keypair(ml_dsa_params *params, uint8_t *pk, uint8_t *sk) {
+  uint8_t seed[SEEDBYTES];
+  /* Get randomness for rho, rhoprime and key */
+  if (!RAND_bytes(seed, SEEDBYTES)) {
+    return -1;
+  }
+  crypto_sign_keypair_internal(params, pk, sk, seed);
+  return 0;
+}
+
+/*************************************************
+* Name:        crypto_sign_signature_internal
+*
+* Description: FIPS 204: Algorithm 7 ML-DSA.Sign_internal.
+*              Computes signature. Internal API.
+*
+* Arguments:   - ml_dsa_params:  parameter struct
 *              - uint8_t *sig:   pointer to output signature (of length CRYPTO_BYTES)
 *              - size_t *siglen: pointer to output length of signature
 *              - uint8_t *m:     pointer to message to be signed
 *              - size_t mlen:    length of message
-*              - uint8_t *ctx:   pointer to context string
-*              - size_t ctxlen:  length of context string
+*              - uint8_t *pre:   pointer to prefix string
+*              - size_t prelen:  length of prefix string
+*              - uint8_t *rnd:   pointer to random seed
 *              - uint8_t *sk:    pointer to bit-packed secret key
 *
 * Returns 0 (success) or -1 (context string too long)
 **************************************************/
-int crypto_sign_signature(ml_dsa_params *params,
-                          uint8_t *sig,
-                          size_t *siglen,
-                          const uint8_t *m,
-                          size_t mlen,
-                          const uint8_t *ctx,
-                          size_t ctxlen,
-                          const uint8_t *sk)
+int crypto_sign_signature_internal(ml_dsa_params *params,
+                                   uint8_t *sig,
+                                   size_t *siglen,
+                                   const uint8_t *m,
+                                   size_t mlen,
+                                   const uint8_t *pre,
+                                   size_t prelen,
+                                   const uint8_t *rnd,
+                                   const uint8_t *sk)
 {
   unsigned int n;
-  uint8_t seedbuf[2*SEEDBYTES + TRBYTES + RNDBYTES + 2*CRHBYTES];
-  uint8_t *rho, *tr, *key, *mu, *rhoprime, *rnd;
+  uint8_t seedbuf[2*SEEDBYTES + TRBYTES + 2*CRHBYTES];
+  uint8_t *rho, *tr, *key, *mu, *rhoprime;
```

Comment:
I'm not sure I understand why we are doing this super convoluted thing with a single buffer and pointers to parts of the buffer. Why aren't `rho, tr, ...` individual buffers?

Reply:
This is upstream code (https://github.com/pq-crystals/dilithium/blob/444cdcc84eb36b66fe27b3a2529ee48f6d8150c2/ref/sign.c#L95-L96).

To aid review I'm trying to not touch upstream code too much, so we can compare against the reference implementation. I agree that a lot of the reference code is convoluted (probably because they are avoiding dependencies). However, I don't think we're at a stage of development in which we want to radically differ from the reference implementation. Unless necessary for security, or compliance to the standard.

If you have a cleaner proposal for this, I'm happy to take it -- I just don't want to be radically changing the upstream reference before you both get to review with no basis to understand these changes.

Reply:
fair enough!